src/python/pants/backend/jvm/tasks/jvm_compile/rsc/rsc_compile.py

```diff
@@ -735,3 +735,21 @@ def _jdk_libs_paths_and_digest(self, hermetic_dist):
   @memoized_method
   def _jdk_libs_abs(self, nonhermetic_dist):
     return nonhermetic_dist.find_libs(self._JDK_LIB_NAMES)
+
+  def _double_check_cache_for_vts(self, vts, zinc_compile_context):
+    # Double check the cache before beginning compilation
+    if self.check_cache(vts):
+      self.context.log.debug(f'Snapshotting results for {vts.target.address.spec}')
+      classpath_entry = self._classpath_for_context(zinc_compile_context)
+      assert classpath_entry.directory_digest is None, f'Classpath entry {classpath_entry} from a double-check is expected to *not* have a digest set!'
+      assert os.path.isfile(classpath_entry.path), f'Classpath entry {classpath_entry} from a double-check should point to an existing file!'
+      relpath = fast_relpath(classpath_entry.path, get_buildroot())
+      classes_dir_snapshot, = self.context._scheduler.capture_snapshots([
+        PathGlobsAndRoot(
+          PathGlobs([relpath]),
+          get_buildroot(),
+        ),
+      ])
+      classpath_entry.set_directory_digest(classes_dir_snapshot.directory_digest)
+      # Re-validate the vts!
+      vts.update()
```

Comment:
Yes, but the `on_success()` kwarg only applies for the `self._default_work_for_vts` method, I think. We could make the `on_success` callback an argument to `self._double_check_cache_for_vts` here as well, but it seemed like for a job specifically for checking the cache that we would always want to do that on success. Do you think it would be good to accept `on_success` and `on_failure` arguments for the double check job too? It would *definitely* make sense, I think, since there's definitely a "success" and "failure" case here.

Reply:
This is the status quo: https://github.com/pantsbuild/pants/blob/33a6d51db91ea1fe69c39117c8878abcb824cdfb/src/python/pants/backend/jvm/tasks/jvm_compile/jvm_compile.py#L941-L944
Because it's doing cache out of band (outside cache manager), I think it should be fine to follow the old pattern for now.